### PR TITLE
feat(python): additional `schema_overrides` param for more ergonomic `DataFrame` init

### PR DESCRIPTION
**Pull request by alexander-beedie**
ref #6209.

This turned into more effort than anticipated... heads-up if you ever want to head into our constructor labyrinth; you may not come back - bring coffee :)

# What it does

As @stinodego accurately pointed out in the linked issue, we do not currently have an ergonomic way to specify specific dtypes for just one or two columns out of a larger set at frame-init time; it tends to be either all or nothing, require the full list of columns, etc - this use-case can be said to _Not Have A Good Story™_. 

The primary suggestion was to add a new `dtypes` param that can apply full _or_ partial schema information, supplementing or overriding anything gleaned from `columns`, and allowing better delineation of column names/order/types. That is what this PR implements (though we have since settled on `schema_overrides`).

While I was at it I found that quite a few of the `pl.from_<xyz>` methods didn't actually support `columns`, though in the main they pass-through to underlying functions that do; these functions have been upgraded to offer both `columns` and `schema_overrides`).

# Example

**Before**: 
_(have to refer to every column to override one)_
```python
df = pl.DataFrame(
    data = {
        "a": [1, 2 ,3],
        "b": [1, 2 ,3],
        "c": [1, 2 ,3],
        "d": [1, 2 ,3],
        "e": [1, 2 ,3],
    },
    columns = ["a", "b", "c", ("d",pl.Int8), "e"],
)
# or: columns = {"a":None, "b":None, "c":None, "d":pl.Int8, "e":None}`
```
^^ Now imagine how bad this gets for frames with 100s of columns ;)
 
**After**: 
_(previous code would still work, but now we have a much cleaner option)_
```python
df = pl.DataFrame(
    data = {
        "a": [1, 2 ,3],
        "b": [1, 2 ,3],
        "c": [1, 2 ,3],
        "d": [1, 2 ,3],
        "e": [1, 2 ,3],
    },
    schema_overrides = {"d": pl.Int8},
)
```

# Bug fixes

Several; carefully walking through the constructors again and adding more unit tests uncovered a number of previously unreported issues. Existing type overrides could potentially be ignored (eg: dataclass/namedtuple input), cause errors, or (in one especially odd case involving pyarrow) reshape the frame `repr` output in an unexpected way. I wouldn't describe any of them as major, but all have now been fixed.

# Next steps

This PR is completely backwards-compatible and additive to what we already have; aside from the bug fixes there are no changes to the current default behaviour, and no need for any changes in existing code. The new param can simply be used to write cleaner code going forward, and it lives happily alongside all of the overloads currently possible via `columns`.

Now that this gives us a more ergonomic way to apply partial schemas we have a better base to build on if there is interest in reworking it further (different param names like "schema", etc); this commit definitely doesn't address everything from the #6209 discussion, but it does take care of the biggest pain point.

I'd suggest closing out the existing issue, and starting a new one that refocuses on the remaining areas of interest?

(Should probably run through the docstrings and add more examples using `schema_overrides` too).

# Also 

_Miscellaneous:_

I know I've seen a request for this somewhere (possibly the Discord - couldn't find it in the open issue list), so I added the negative counterparts of our existing assert functions as they made sense for a couple of the updated tests:

* `assert_frame_not_equal`
* `assert_series_not_equal`

### CODE REVIEW

**Comment by stinodego** [comment]
Thanks! I'll take a look tomorrow :)

File: py-polars/tests/unit/test_df.py [hunk 1]
```diff
@@ -2189,11 +2231,15 @@ def test_product() -> None:
...
+        },
+        dtypes={
+            "int": pl.UInt16,
+            "flt": pl.Float32,
+        },
     )
     out = df.product()
     expected = pl.DataFrame({"int": [6], "flt": [-108.0], "bool_0": [0], "bool_1": [1]})
-    assert out.frame_equal(expected)
+    assert_frame_not_equal(out, expected)
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/test_df.py:2242]: Just out of interest, why not construct `expected` explicitly (like with `.cast`, so it goes down a different path than it would with the `dtypes` constructor) and compare against that? There's an infinite number of arguments that would satisfy `assert_frame_not_equal` 😄

Comment (alexander-beedie) [reply py-polars/tests/unit/test_df.py:2242]: Ah; because I probably meant to write this; good catch, will update ;)
```python
    assert_frame_not_equal(out, expected, check_dtype=True)
    assert_frame_equal(out, expected, check_dtype=False)
```

**Comment by alexander-beedie** [comment]
(Just updated further, with extra tests and `dtypes` support for `pl.from_pandas` and `pl.from_numpy`)

**Comment by stinodego** [comment]
@alexander-beedie I just added a comment to the original issue with some more thoughts and an alternative suggestion. Wondering what you think.

I just want to make sure we get this right! I want to avoid adding additional options to the constructor of our 'main' object that will be hard to deprecate/change later if we're not 100% sure.

**Comment by alexander-beedie** [comment]
> @alexander-beedie I just added a comment to the original issue with some more thoughts and an alternative suggestion. Wondering what you think.

I have made my counter-proposal, heh. I agree that we should rename the params, and also enforce a better separation of responsibilities - but I disagree that we should be overloading a single parameter with all of that. But I think we can actually get a _"best of both worlds"_ result here - see what you think ;)

**Comment by alexander-beedie** [comment]
@stinodego: should be good to go. I might actually have some connectivity in the air, so perhaps I'll try and set a "highest altitude" polars commit record later... ;)

**Comment by ritchie46** [comment]
> @stinodego: should be good to go. I might actually have some connectivity in the air, so perhaps I'll try and set a "highest altitude" polars commit record later, with the `columns` deprecation... ;)

LOL.. the polars sky club🚀😂

File: py-polars/polars/convert.py [hunk 2]
```diff
@@ -50,14 +55,17 @@ def from_dict(
     └─────┴─────┘
 
     """
-    return DataFrame._from_dict(data=data, columns=columns)
+    return DataFrame._from_dict(
+        data=data, columns=columns, schema_overrides=schema_overrides
+    )
 
 
+@deprecated_alias(schema="dtypes")
```
Comment (stinodego) [inline py-polars/polars/convert.py:63]:
```suggestion
@deprecated_alias(schema="schema_overrides")
```

File: py-polars/polars/convert.py [hunk 3]
```diff
@@ -50,14 +55,17 @@ def from_dict(
...
 def from_dicts(
     dicts: Sequence[dict[str, Any]],
     infer_schema_length: int | None = N_INFER_DEFAULT,
     *,
-    schema: Schema | None = None,
+    schema_overrides: Schema | None = None,
```
Comment (stinodego) [inline py-polars/polars/convert.py:68]: Functionally, this parameter seems to be equivalent to `schema_overrides`. But we probably will want to add a `schema` parameter too, right?

We may want to do that in a separate PR, but I see that elsewhere you have added both.

File: py-polars/polars/convert.py [hunk 4]
```diff
@@ -120,12 +130,15 @@ def from_dicts(
...
-    return DataFrame._from_dicts(dicts, infer_schema_length, schema)
+    return DataFrame._from_dicts(
+        dicts, infer_schema_length, schema_overrides=schema_overrides
+    )
 
 
 def from_records(
     data: Sequence[Sequence[Any]],
     columns: Sequence[str] | None = None,
+    schema_overrides: Schema | None = None,
```
Comment (stinodego) [inline py-polars/polars/convert.py:141]: Technically, adding this as the third parameter (which makes sense) is a breaking change, in case people were specifying `orient` as a positional parameter.

Not sure if we should worry about that, but if we want to do it nicely, we should add it as the last parameter and then move it to the correct position in the next breaking release.

This same logic affects a lot of the changes in this PR.

File: py-polars/polars/convert.py [hunk 5]
```diff
@@ -219,11 +243,16 @@ def from_numpy(
...
-    return DataFrame._from_numpy(data, columns=columns, orient=orient)
+    return DataFrame._from_numpy(
+        data, columns=columns, orient=orient, schema_overrides=schema_overrides
+    )
 
 
 def from_arrow(
-    a: pa.Table | pa.Array | pa.ChunkedArray, rechunk: bool = True
+    a: pa.Table | pa.Array | pa.ChunkedArray,
+    columns: Sequence[str] | None = None,
```
Comment (stinodego) [inline py-polars/polars/convert.py:253]: Since we're adding a new parameter, we should give it the right name straight away.
```suggestion
    schema: Sequence[str] | None = None,
```

File: py-polars/tests/unit/test_df.py [hunk 6]
```diff
@@ -223,26 +231,39 @@ def test_from_arrow() -> None:
...
-    assert df.columns == ["a", "b"]
-    assert df.rows() == [(1, 3), (2, 4)]
-
-    # expected error
-    mismatched_schema = {"x": pl.UInt8, "b": pl.UInt32}
-    with pytest.raises(ValueError):
-        pl.DataFrame({"b": [3, 4], "a": [1, 2]}, columns=mismatched_schema)
+    data = {"b": [3, 4], "a": [1, 2]}
+    for df in (
+        pl.DataFrame(data, columns={"a": pl.UInt8, "b": pl.UInt32}),
```
Comment (stinodego) [inline py-polars/tests/unit/test_df.py:248]:
```suggestion
        pl.DataFrame(data, columns=schema),
```

**Review by stinodego — COMMENTED** [review]
Just a few comments. Nice improvement!

**Comment by alexander-beedie** [comment]
> Just a few comments. Nice improvement!

Great, thanks. You have the eye of the tiger ;)

We're on the runway and moving now - if there's connectivity at 35000 feet I'll take care of it! (And if not, then tomorrow in Tokyo ;)

File: py-polars/polars/convert.py (hunk 3 above)
Comment (alexander-beedie) [reply py-polars/polars/convert.py:68]: Yep; I'll add `schema` in the next round (changing `columns`) to bring it up to equivalence.

File: py-polars/polars/convert.py (hunk 2 above)
Comment (alexander-beedie) [reply py-polars/polars/convert.py:63]: Good catch

File: py-polars/polars/convert.py (hunk 5 above)
Comment (alexander-beedie) [reply py-polars/polars/convert.py]: Done. I've also specced a new docstring that (hopefully ;) better explains the `schema` param. Using this function as a testbed for the; new explanation see what you think?

File: py-polars/polars/convert.py (hunk 4 above)
Comment (alexander-beedie) [reply py-polars/polars/convert.py]: Ok, I've shifted them to the end; we can pair them back up for 0.16 - next PR will be a straight `columns => schema` update in conjunction with `deprecated_alias`, and then a follow-up can move the params together (with a breaking notification on the commit).

**Comment by alexander-beedie** [comment]
@stinodego: All remaining comments addressed; also changed the name of the `ColumnsType` alias to `SchemaDefinition` to better describe it, in anticipation of the imminent `columns => schema` update.

File: py-polars/polars/convert.py (hunk 4 above)
Comment (stinodego) [reply py-polars/polars/convert.py]: I see that the `DataFrame` class constructor still has `schema_overrides` before `orient`. That should really be adjusted, as there must be people initiating their DataFrames with something like `pl.DataFrame([[1,2],[3,4]], ['a', 'b'], 'row')`, which this will break.

Comment (ritchie46) [reply py-polars/polars/convert.py]: Shall we make it keyword only?

Comment (alexander-beedie) [reply py-polars/polars/convert.py]: > I see that the `DataFrame` class constructor still has `schema_overrides` before `orient`. That should really be adjusted, as there must be people initiating their DataFrames with something like `pl.DataFrame([[1,2],[3,4]], ['a', 'b'], 'row')`, which this will break.

Doh! Missed it... will fix ASAP; `kwargs`-only doesn't sound like a bad idea?

Comment (stinodego) [reply py-polars/polars/convert.py]: Exactly, we should probably add a few "kwarg only" stars in the next breaking release. You could already make this one kwarg-only.

Comment (alexander-beedie) [reply py-polars/polars/convert.py]: Ok, have updated accordingly - along with making the `infer_schema_lengths` param kwargs-only too. That param did technically sneak out 0.15.15, but it feels more like a kwargs-only param and nobody will be using it yet, so...

**Review by stinodego — APPROVED** [review]
Looks like we can ship it!

**Comment by alexander-beedie** [comment]
@ritchie46: ready to roll :)

**Comment by ritchie46** [comment]
Alright! Thanks @alexander-beedie and @stinodego!

**Comment by alexander-beedie** [comment]
> Alright! Thanks @alexander-beedie and @stinodego!

Great! Will start on the second phase update shortly... ;)
